codegen/src/test/resources/software/amazon/awssdk/codegen/poet/rules/endpoint-auth-scheme-interceptor.java

```diff
@@ -0,0 +1,54 @@
+package software.amazon.awssdk.services.query.rules.internal;
+
+import java.util.List;
+import java.util.function.Supplier;
+import software.amazon.awssdk.annotations.Generated;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.auth.signer.Aws4Signer;
+import software.amazon.awssdk.auth.signer.SignerLoader;
+import software.amazon.awssdk.awscore.AwsRequest;
+import software.amazon.awssdk.awscore.rules.AwsEndpointAttribute;
+import software.amazon.awssdk.awscore.rules.AwsEndpointProviderUtils;
+import software.amazon.awssdk.awscore.rules.EndpointAuthScheme;
+import software.amazon.awssdk.awscore.util.SignerOverrideUtils;
+import software.amazon.awssdk.core.SdkRequest;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.core.interceptor.Context;
+import software.amazon.awssdk.core.interceptor.ExecutionAttributes;
+import software.amazon.awssdk.core.interceptor.ExecutionInterceptor;
+import software.amazon.awssdk.core.interceptor.SdkInternalExecutionAttribute;
+import software.amazon.awssdk.core.rules.model.Endpoint;
+import software.amazon.awssdk.core.signer.Signer;
+
+@Generated("software.amazon.awssdk:codegen")
+@SdkInternalApi
+public final class QueryEndpointAuthSchemeInterceptor implements ExecutionInterceptor {
+    @Override
+    public SdkRequest modifyRequest(Context.ModifyRequest context, ExecutionAttributes executionAttributes) {
+        Endpoint resolvedEndpoint = executionAttributes.getAttribute(SdkInternalExecutionAttribute.RESOLVED_ENDPOINT);
+        AwsRequest request = (AwsRequest) context.request();
+        if (resolvedEndpoint.headers() != null) {
+            request = AwsEndpointProviderUtils.addHeaders(request, resolvedEndpoint.headers());
+        }
```

Comment:
If all headers aren't concerned with auth schemes this is an odd place to put this logic.

Reply:
Yeah I think so too. Is this a blocker? If not, I'll just keep it here to avoid adding another interceptor

Reply:
Does it need to be an interceptor though? I know this part of the code base will continue to go through a lot of changes not only with endpoints but also with other core refactoring that's planned so I guess we can live with it if it's short/medium term. I'm just a bit concerned that we're dropping pieces of mutating logic in places where it may be hard to find them.

Reply:
The only other place it could go is the `AwsExecutionContextBuilder`, but doing so won't get rid of the interceptor. It's easier to reason about when they're all in the interceptors, as opposed to splitting them up between within the interceptors and in core IMO

Reply:
Let's leave it then and see what the future holds :-)